jekyll/_cci2/writing-shell-scripts.md

```diff
@@ -0,0 +1,88 @@
+---
+layout: classic-docs
+title: "Writing Shell Scripts"
+short-title: "Writing Shell Scripts"
+description: "Best practices for writing shell scripts for use in CircleCI configuration"
+categories: [getting-started]
+order: 10
+---
+
+*[Basics]({{ site.baseurl }}/2.0/basics/) > Writing Shell Scripts*
+
+## Overview
+
+Configuring CircleCI often requires
+writing shell scripts.
+While shell scripting can grant finer control over your build,
+it is a subtle art
+that can produce equally subtle errors.
+You can avoid many of these errors
+by reviewing the best practices
+explained below.
+
+### Use ShellCheck
+
+[ShellCheck](https://github.com/koalaman/shellcheck) is a shell script static analysis tool
+that saves you from yourself,
+regardless of your level of shell scripting experience.
+
+ShellCheck works best with CircleCI
+when you add it as a separate job in your `.circleci/config.yml` file.
+This allows you
+to run the `shellcheck` job in parallel with other jobs in a workflow.
+
+```yaml
+version: 2
+jobs:
+  shellcheck:
+    docker:
+      - image: nlknguyen/alpine-shellcheck:v0.4.6
+    steps:
+      - checkout
+      - run:
+          name: Check Scripts
+          command: |
+            find . -type f -name '*.sh' | wc -l
+            find . -type f -name '*.sh' | xargs shellcheck --external-sources
+  build-job:
+    ...
+
+workflows:
+  version: 2
+  workflow:
+    jobs:
+      - shellcheck
+      - build-job:
+          requires:
+            - shellcheck
+          filters:
+            branches:
+              only: master
+```
+
+### Set Error Flags
+
+There are several error flags
+you can set
+to automatically exit scripts
+when unfavorable conditions occur.
+As a best practice,
+add the following flags at the beginning of each script
+to protect yourself from tricky errors.
+
+```bash
+#!/usr/bin/env bash
+
+# Exit script if you try to use an uninitialized variable.
+set -o nounset
+
+# Exit script if a statement returns a non-true return value.
+set -o errexit
+
+# Use the error status of the first failure, rather than that of the last item in a pipeline.
+set -o pipefail
+```
+
```

Comment:
add the parts about env vars / secrets and how to prevent env var expansion. If there are some tickets about this, it would help to know how customers describe these problems when they have them and mold the content around that thinking/those words/questions.

Reply:
Can you clarify what parts about env vars & secrets you mean? A link to another document or ticket would be very helpful here.

Reply:
Ah, I found what you meant: you're referring to Tom's comment in the attached JIRA ticket. I didn't document that because I believe it was related to the `set -o xtrace` command (which is not added in this PR).

Reply:
Can you double-check that, it seems like the most important part of this documentation to me. That is, what can we say about shell expansion of env var secrets so they don't get displayed in build output? Examples where our users have gone wrong?